__repr__ method for Novel

kmerrill18 opened the pull request:
Overrides python's formal representation so that a novel as represented as its filename without the extension, so author underscore title word. I thought I committed other changes in other pull requests, but I'm seeing other commits here, so an explanation for those just in case. Those include additions to the pronoun frequency analysis code to include another way of looking at subject and object pronouns and a markdown write-up of that data.

equationcrunchor commented:
Aren’t filenames now going to be ID’s?

srisi commented on gender_novels/novel.py:
```diff
@@ -119,6 +119,30 @@ def __str__(self):
         name = self.filename[0:len(self.filename)-4]
         return name
 
+    def __repr__(self):
```
For the time being, it looks like __repr__ returns exactly the same thing as __str__ in which case, you could be lazy 
and make the code just 
```
return self.__str__()
```
Now, my understanding is that that would be perfectly fine but also that the expectation for repr is that it should return a bit more useful information for a developer. Though I should say that I haven't written a __repr__ function before. One possible way of doing that would be to have repr also show that you are dealing with a Novel object: 
```
        name = self.filename[0:len(self.filename) - 4]
        return f'Novel ({name})'
```

mscuthbert replied:
Sometimes people put this in angle brackets, so

```
return f'<Novel ({name})>'
```
would match convention.  Generally we either put `<ClassName someinfo>` or if it's possible to recreate the object easily then `Novel(name)` (but I don't think this is the case).

srisi commented on gender_novels/analysis/gender_pronoun_freq_analysis.py:
```diff
@@ -62,6 +62,17 @@ def subject_vs_object_pronoun_freqs(corp):
 
     return result_tuple
 
+def sub_pronouns_gender_comp(corp):
```
This function needs a bit of work before we can merge it. 
- name: `subject_pronouns_gender_comparison` is longer but more immediately descriptive
- generality: can you write the function so that it accepts something like a `subject_gender` to calculate? That way, it can be used to calculate relative male and female subject frequencies without changing any of the code.
- documentation: what does it do?
- doc test: how to use it and does it work?

kmerrill18 replied:
I'm confused on what you are asking here. I understand the documentation bit - sorry, that was an oversight. Will do. What do you mean by:

generality: can you write the function so that it accepts something like a subject_gender to calculate? That way, it can be used to calculate relative male and female subject frequencies without changing any of the code.

There's already a function above it called subject_vs_object_pronoun_freqs that calculates the percent of pronouns of one gender that are subject pronouns. This function looks at the total number of subject pronouns and calculates the portion of those that are male and female. This particular function is highly biased to the proportion of total pronouns of each gender, and I realized that when I wrote it. I just wanted to make sure I wasn't overlooking anything. I.e., if this function returned significantly different numbers than the function that gets proportions for total pronouns, there would be something different about subject pronouns than pronouns in general. That was my thinking anyway.

srisi replied:
Sure. I should have been clearer. The function is very useful. However, it always returns the frequency of female subjects and it would be nice if we could also directly get the frequency of male subjects. It's a mathematically simple operation but there is nothing in the function that would tell me that I'm getting back the frequency of female subjects rather than both. 
So the change could be: 
```
def sub_pronouns_gender_comp(corp, subject_gender):

    relative_frequeny_of_female_subjects = {}
    relative_frequency_of_male_subjects = {}

    for book in corp.novels:
        he = book.get_word_freq('he')
        she = book.get_word_freq('she')
        
        relative_frequency_of_female_subjects[book] = (she) / (he + she)
        relative_frequency_of_male_subjects[book] = (he) / (he + she)
    
    if subject_gender == 'female':
        return relative_frequency_of_female_subjects
    if subject_gender == 'male':
        return relative_frequency_of_male_subjects
```

mscuthbert commented:
Great work!  Tiny bit of changes then this'll be AWESOME to have!  :-)

kmerrill18 commented:
I added the missing docstrings. I did not add doctests due to the issue I have described earlier of not having a corpus smaller than ~100 novels to test on and not wanting to take many seconds to a few minutes to run tests. I fixed __repr__ function per suggestions to fit with conventions, and parameter subject_gender was added to subject_pronouns_gender_comparison. The function now raises a ValueError if the subject_gender passed in is not 'male' or 'female'.

srisi commented:
Thank you!
